File: src/main/java/org/lizlooney/mandlebrot/Mandlebrot.java
/*
 * Copyright 2023 Liz Looney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.lizlooney.mandlebrot;

public class Mandlebrot {
  public static final int MAX_VALUE = 1000;

  private final String s;
  private final boolean useNativeCode;
  private final int numThreads;
  private final int sizeInPixels;
  private final double size;
  private final double pixelsPerUnit;
  private final double aMin;
  private final double bMin;
  private final Object valuesLock = new Object();
  private final int[] values;

  public Mandlebrot(boolean useNativeCode, int numThreads,
      int sizeInPixels, double aCenter, double bCenter, double size) {
    this.s = "Center: (" + formatDouble(aCenter) + ", " + formatDouble(bCenter) + ") width/height: " + formatDouble(size);

    this.useNativeCode = useNativeCode;
    this.numThreads = numThreads;
    this.sizeInPixels = sizeInPixels;
    this.size = size;
    pixelsPerUnit = sizeInPixels / size;
    aMin = aCenter - size / 2;
    bMin = bCenter - size / 2;
    values = new int[sizeInPixels * sizeInPixels];

    long startTime = System.currentTimeMillis();
    calculatePixelValues();
    long endTime = System.currentTimeMillis();
    System.out.println("HeyLiz - useNativeCode is " + useNativeCode);
    System.out.println("HeyLiz - numThreads is " + numThreads);
    System.out.println("HeyLiz - calculatePixelValues took " + ((endTime - startTime) / 1000.0) + " seconds");
  }

  private static String formatDouble(double d) {
    String s = String.format("%f", d);
    if (s.contains(".")) {
      while (s.endsWith("0")) {
        s = s.substring(0, s.length() - 1);
      }
      if (s.endsWith(".")) {
        s = s.substring(0, s.length() - 1);
      }
    }
    return s;
  }

  public Mandlebrot panZoom(int x, int y, double zoomFactor) {
    double cA = aMin + x / pixelsPerUnit;
    double cB = bMin + y / pixelsPerUnit;
    return new Mandlebrot(useNativeCode, numThreads, sizeInPixels, cA, cB, size * zoomFactor);
  }

  public String toString() {
    return s;
  }

  private static native void calculatePixelValuesNative(int[] values, int numThreads, double aMin, double bMin, double pixelsPerUnit, int sizeInPixels, int maxValue);

  private void calculatePixelValues() {
    if (useNativeCode) {
      calculatePixelValuesNative(values, numThreads, aMin, bMin, pixelsPerUnit, sizeInPixels, MAX_VALUE);
      return;
    }

    Thread[] threads = new Thread[numThreads];
    for (int i = 0; i < numThreads; i++) {
      final int threadNumber = i;
      threads[i] = new Thread(() -> calculateValuesForThread(threadNumber));
      threads[i].start();
    }

    for (int i = 0; i < numThreads; i++) {
      try {
        threads[i].join();
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }

  private void calculateValuesForThread(int threadNumber) {
    int i = 0;
    for (int y = 0; y < sizeInPixels; y++) {
      for (int x = 0; x < sizeInPixels; x++) {
        if (numThreads == 1 || i % numThreads == threadNumber) {
          double cA = aMin + x / pixelsPerUnit;
          double cB = bMin + y / pixelsPerUnit;
          int value = calculateValue(cA, cB);

          synchronized (valuesLock) {
            values[i] = value;
          }
        }
        i++;
      }
    }
  }

  private static native int calculateValueNative(double cA, double cB, int maxValue);

  private int calculateValue(double cA, double cB) {
    if (useNativeCode) {
      return calculateValueNative(cA, cB, MAX_VALUE);
    }

    double zA = cA;
    double zB = cB;
    for (int i = 0; i <= MAX_VALUE; i++) {
      double zAzA = zA * zA;
      double zBzB = zB * zB;
      if (zAzA + zBzB >= 4) {
        return i;
      }
      double nextA = zAzA - zBzB + cA;
      zB = 2 * zA * zB + cB;
      zA = nextA;
    }
    return Integer.MAX_VALUE;
  }

  public interface Visitor {
    void visit(int x, int y, int value);
  }

  public void accept(Visitor visitor) {
    synchronized (valuesLock) {
      int i = 0;
      for (int y = 0; y < sizeInPixels; y++) {
        for (int x = 0; x < sizeInPixels; x++) {
          visitor.visit(x, y, values[i]);
          i++;
        }
      }
    }
  }
}

File: src/main/python/mandlebrot.py
"""Module mandlebrot calculates pixel values for mandlebrot fractals"""

# Copyright 2023 Liz Looney
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Final

MAX_VALUE: Final[int] = 1000

class Mandlebrot():
    """Class that does calculations for Mandlebrot fractals."""

    size_in_pixels: int
    pixels_per_unit: float
    a_min: float
    b_min: float
    # HeyLiz values: int[]

    def __init__(self, size_in_pixels: int, a_center: float, b_center: float, size: float):
        self.size_in_pixels = size_in_pixels
        self.size = size
        self.pixels_per_unit = size_in_pixels // size
        self.a_min = a_center - size / 2
        self.b_min = b_center - size / 2
        self.values = [ 0 for i in range(size_in_pixels * size_in_pixels) ]
        self.calculate_pixel_values()

    def calculate_pixel_values(self):
        """Function that calculates the values for all pixels"""
        i = 0
        for y_coord in range(self.size_in_pixels):
            for x_coord in range(self.size_in_pixels):
                complex_number = complex(
                    self.a_min + x_coord / self.pixels_per_unit,
                    self.b_min + y_coord / self.pixels_per_unit)
                self.values[i] = Mandlebrot.calculate_value(complex_number)
                i += 1

    @staticmethod
    def calculate_value(complex_number: complex):
        """Function that calculates the value for the given complex number"""
        z_value = complex(complex_number)
        for i in range(MAX_VALUE):
            z_squared = z_value * z_value
            if abs(z_squared) >= 4:
                return i
            z_value = z_squared + complex_number
        return MAX_VALUE

    def show_values(self):
        """Function that shows the values as a grid of printable ascii characters"""
        i = 0
        for _ in range(self.size_in_pixels):
            for _ in range(self.size_in_pixels):
                value = 0x20 + self.values[i] % (0x7E - 0x20)
                print(chr(value), end='')
                i += 1
            print()


if __name__ == '__main__':
    mandlebrot = Mandlebrot(100, 0, 0, 4)
    mandlebrot.show_values()
